File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/replay/ReplayRouteLocationEngine.java
package com.mapbox.services.android.navigation.v5.location.replay;

import android.annotation.SuppressLint;
import android.app.PendingIntent;
import android.location.Location;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mapbox.android.core.location.LocationEngine;
import com.mapbox.android.core.location.LocationEngineCallback;
import com.mapbox.android.core.location.LocationEngineRequest;
import com.mapbox.android.core.location.LocationEngineResult;
import com.mapbox.api.directions.v5.models.DirectionsRoute;
import com.mapbox.geojson.LineString;
import com.mapbox.geojson.Point;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;


public class ReplayRouteLocationEngine implements LocationEngine, Runnable {
    private static final int HEAD = 0;
    private static final int MOCKED_POINTS_LEFT_THRESHOLD = 5;
    private static final int ONE_SECOND_IN_MILLISECONDS = 1000;
    private static final int FORTY_FIVE_KM_PER_HOUR = 45;
    private static final int DEFAULT_SPEED = FORTY_FIVE_KM_PER_HOUR;
    private static final int ONE_SECOND = 1;
    private static final int DEFAULT_DELAY = ONE_SECOND;
    private static final int DO_NOT_DELAY = 0;
    private static final int ZERO = 0;
    private static final String SPEED_MUST_BE_GREATER_THAN_ZERO_KM_H = "Speed must be greater than 0 km/h.";
    private static final String DELAY_MUST_BE_GREATER_THAN_ZERO_SECONDS = "Delay must be greater than 0 seconds.";
    private static final String REPLAY_ROUTE = "ReplayRouteLocation";
    private ReplayRouteLocationConverter converter;
    private int speed = DEFAULT_SPEED;
    private int delay = DEFAULT_DELAY;
    private Handler handler;
    private List<Location> mockedLocations;
    private ReplayLocationDispatcher dispatcher;
    private Location lastLocation = null;
    private CopyOnWriteArrayList<LocationEngineCallback<LocationEngineResult>> callbackList = new CopyOnWriteArrayList<>();
    private final ReplayLocationListener replayLocationListener = new ReplayLocationListener() {
        @Override
        public void onLocationReplay(Location location) {
            lastLocation = location;
            for (LocationEngineCallback<LocationEngineResult> result : callbackList) {
                result.onSuccess(LocationEngineResult.create(location));
            }
            lastLocation = location;
            if (!mockedLocations.isEmpty()) {
                mockedLocations.remove(HEAD);
            }
        }
    };

    public ReplayRouteLocationEngine() {
        this.handler = new Handler();
    }

    @SuppressLint("MissingPermission")
    public void assign(DirectionsRoute route) {
        start(route);
    }

    @SuppressLint("MissingPermission")
    public void moveTo(Point point) {
        Location lastLocation = getLastLocation();
        if (lastLocation == null) {
            return;
        }

        startRoute(point, lastLocation);
    }

    public void assignLastLocation(Point currentPosition) {
        initializeLastLocation();
        lastLocation.setLongitude(currentPosition.longitude());
        lastLocation.setLatitude(currentPosition.latitude());
    }

    public void updateSpeed(int customSpeedInKmPerHour) {
        if (customSpeedInKmPerHour <= 0) {
            throw new IllegalArgumentException(SPEED_MUST_BE_GREATER_THAN_ZERO_KM_H);
        }
        this.speed = customSpeedInKmPerHour;
    }

    public void updateDelay(int customDelayInSeconds) {
        if (customDelayInSeconds <= 0) {
            throw new IllegalArgumentException(DELAY_MUST_BE_GREATER_THAN_ZERO_SECONDS);
        }
        this.delay = customDelayInSeconds;
    }

    @Override
    public void run() {
        List<Location> nextMockedLocations = converter.toLocations();
        if (nextMockedLocations.isEmpty()) {
            handler.removeCallbacks(this);
            return;
        }
        dispatcher.add(nextMockedLocations);
        mockedLocations.addAll(nextMockedLocations);
        scheduleNextDispatch();
    }

    private void start(DirectionsRoute route) {
        handler.removeCallbacks(this);
        converter = new ReplayRouteLocationConverter(route, speed, delay);
        converter.initializeTime();
        mockedLocations = converter.toLocations();
        dispatcher = obtainDispatcher();
        dispatcher.run();
        scheduleNextDispatch();
    }

    private ReplayLocationDispatcher obtainDispatcher() {
        if (dispatcher != null) {
            dispatcher.stop();
            dispatcher.removeReplayLocationListener(replayLocationListener);
        }
        dispatcher = new ReplayLocationDispatcher(mockedLocations);
        dispatcher.addReplayLocationListener(replayLocationListener);

        return dispatcher;
    }

    private void startRoute(Point point, Location lastLocation) {
        handler.removeCallbacks(this);
        converter.updateSpeed(speed);
        converter.updateDelay(delay);
        converter.initializeTime();
        LineString route = obtainRoute(point, lastLocation);
        mockedLocations = converter.calculateMockLocations(converter.sliceRoute(route));
        dispatcher = obtainDispatcher();
        dispatcher.run();
    }

    @NonNull
    private LineString obtainRoute(Point point, Location lastLocation) {
        List<Point> pointList = new ArrayList<>();
        pointList.add(Point.fromLngLat(lastLocation.getLongitude(), lastLocation.getLatitude()));
        pointList.add(point);
        return LineString.fromLngLats(pointList);
    }

    private void scheduleNextDispatch() {
        int currentMockedPoints = mockedLocations.size();
        if (currentMockedPoints == ZERO) {
            handler.postDelayed(this, DO_NOT_DELAY);
        } else if (currentMockedPoints <= MOCKED_POINTS_LEFT_THRESHOLD) {
            handler.postDelayed(this, ONE_SECOND_IN_MILLISECONDS);
        } else {
            handler.postDelayed(this, (currentMockedPoints - MOCKED_POINTS_LEFT_THRESHOLD) * ONE_SECOND_IN_MILLISECONDS);
        }
    }

    private void initializeLastLocation() {
        if (lastLocation == null) {
            lastLocation = new Location(REPLAY_ROUTE);
        }
    }

    @SuppressLint("MissingPermission")
    @Nullable
    public Location getLastLocation() {
        return lastLocation;
    }

    public void onStop() {
        if (dispatcher != null) {
            dispatcher.stop();
        }
        handler.removeCallbacks(this);
        for (LocationEngineCallback<LocationEngineResult> callback : callbackList) {
            callbackList.remove(callback);
        }
        if (dispatcher != null) {
            dispatcher.removeReplayLocationListener(replayLocationListener);
        }
    }

    @Override
    public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
        if (lastLocation != null) {
            callback.onSuccess(LocationEngineResult.create(lastLocation));
        } else {
            callback.onFailure(new Exception("No last location"));
        }
    }

    @Override
    public void requestLocationUpdates(@NonNull LocationEngineRequest request, @NonNull LocationEngineCallback<LocationEngineResult> callback, @Nullable Looper looper) throws SecurityException {
        callbackList.add(callback);
    }

    @Override
    public void requestLocationUpdates(@NonNull LocationEngineRequest request, PendingIntent pendingIntent) throws SecurityException {
    }

    @Override
    public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
        callbackList.remove(callback);
    }

    @Override
    public void removeLocationUpdates(PendingIntent pendingIntent) {
    }
}

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/location/ReplayRouteLocationEngine.kt
package com.mapbox.navigation.core.location

import android.app.PendingIntent
import android.location.Location
import android.os.Handler
import android.os.Looper
import com.mapbox.android.core.location.LocationEngine
import com.mapbox.android.core.location.LocationEngineCallback
import com.mapbox.android.core.location.LocationEngineRequest
import com.mapbox.android.core.location.LocationEngineResult
import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.geojson.LineString
import com.mapbox.geojson.Point
import com.mapbox.navigation.base.location.ReplayLocationConverter
import com.mapbox.navigation.base.logger.Logger
import com.mapbox.navigation.base.logger.model.Message
import com.mapbox.navigation.core.location.replay.ReplayLocationDispatcher
import com.mapbox.navigation.core.location.replay.ReplayRouteLocationConverter
import com.mapbox.navigation.core.location.replay.ReplayRouteLocationListener
import java.util.ArrayList

class ReplayRouteLocationEngine(
    val logger: Logger? = null,
    private val converter: ReplayLocationConverter
) : LocationEngine, Runnable {

    private var speed = DEFAULT_SPEED
    private var delay = DEFAULT_DELAY
    private val handler: Handler = Handler()
    private lateinit var mockedLocations: MutableList<Location>
    private lateinit var dispatcher: ReplayLocationDispatcher
    private lateinit var replayLocationListener: ReplayRouteLocationListener
    private lateinit var lastLocation: Location
    private var route: DirectionsRoute? = null
    private var point: Point? = null

    constructor(logger: Logger? = null) : this(
        logger,
        ReplayRouteLocationConverter(
            DEFAULT_SPEED,
            DEFAULT_DELAY
        )
    )

    companion object {
        private const val HEAD = 0
        private const val MOCKED_POINTS_LEFT_THRESHOLD = 5
        private const val ONE_SECOND_IN_MILLISECONDS = 1000
        private const val FORTY_FIVE_KM_PER_HOUR = 45
        private const val DEFAULT_SPEED = FORTY_FIVE_KM_PER_HOUR
        private const val ONE_SECOND = 1
        private const val DEFAULT_DELAY = ONE_SECOND
        private const val DO_NOT_DELAY = 0
        private const val ZERO = 0
        private const val SPEED_MUST_BE_GREATER_THAN_ZERO_KM_H =
            "Speed must be greater than 0 km/h."
        private const val DELAY_MUST_BE_GREATER_THAN_ZERO_SECONDS =
            "Delay must be greater than 0 seconds."
        private const val REPLAY_ROUTE =
            "com.mapbox.services.android.navigation.v5.location.replay.ReplayRouteLocationEngine"
    }

    fun assign(route: DirectionsRoute) {
        this.route = route
        this.point = null
    }

    fun moveTo(point: Point) {
        this.point = point
        this.route = null
    }

    fun assignLastLocation(currentPosition: Point) {
        initializeLastLocation()
        lastLocation.longitude = currentPosition.longitude()
        lastLocation.latitude = currentPosition.latitude()
    }

    fun updateSpeed(customSpeedInKmPerHour: Int) {
        require(customSpeedInKmPerHour > 0) { SPEED_MUST_BE_GREATER_THAN_ZERO_KM_H }
        this.speed = customSpeedInKmPerHour
    }

    fun updateDelay(customDelayInSeconds: Int) {
        require(customDelayInSeconds > 0) { DELAY_MUST_BE_GREATER_THAN_ZERO_SECONDS }
        this.delay = customDelayInSeconds
    }

    override fun run() {
        var nextMockedLocations = converter.toLocations()
        if (nextMockedLocations.isEmpty()) {
            if (converter.isMultiLegRoute) {
                nextMockedLocations = converter.toLocations()
            } else {
                handler.removeCallbacks(this)
                return
            }
        }
        dispatcher.add(nextMockedLocations)
        mockedLocations.addAll(nextMockedLocations)
        scheduleNextDispatch()
    }

    @Throws(SecurityException::class)
    override fun getLastLocation(callback: LocationEngineCallback<LocationEngineResult>) {
        callback.onSuccess(LocationEngineResult.create(lastLocation))
    }

    @Throws(SecurityException::class)
    override fun requestLocationUpdates(
        request: LocationEngineRequest,
        callback: LocationEngineCallback<LocationEngineResult>,
        looper: Looper?
    ) {
        beginReplayWith(callback)
    }

    @Throws(SecurityException::class)
    override fun requestLocationUpdates(
        request: LocationEngineRequest,
        pendingIntent: PendingIntent
    ) {
        logger?.e(msg = Message("ReplayEngine does not support PendingIntent."))
    }

    override fun removeLocationUpdates(callback: LocationEngineCallback<LocationEngineResult>) {
        deactivate()
    }

    override fun removeLocationUpdates(pendingIntent: PendingIntent) {
        logger?.e(msg = Message("ReplayEngine does not support PendingIntent."))
    }

    internal fun updateLastLocation(lastLocation: Location) {
        this.lastLocation = lastLocation
    }

    internal fun removeLastMockedLocation() {
        if (mockedLocations.isNotEmpty()) {
            mockedLocations.removeAt(HEAD)
        }
    }

    private fun deactivate() {
        if (::dispatcher.isInitialized) {
            dispatcher.stop()
        }
        handler.removeCallbacks(this)
    }

    private fun start(
        route: DirectionsRoute,
        callback: LocationEngineCallback<LocationEngineResult>
    ) {
        handler.removeCallbacks(this)
        converter.initializeTime()
        mockedLocations = converter.toLocations().toMutableList()
        dispatcher = obtainDispatcher(callback)
        dispatcher.run()
        scheduleNextDispatch()
    }

    private fun obtainDispatcher(callback: LocationEngineCallback<LocationEngineResult>): ReplayLocationDispatcher {
        if (::dispatcher.isInitialized && ::replayLocationListener.isInitialized) {
            dispatcher.stop()
            dispatcher.removeReplayLocationListener(replayLocationListener)
        }
        dispatcher =
            ReplayLocationDispatcher(
                mockedLocations
            )
        replayLocationListener =
            ReplayRouteLocationListener(
                this,
                callback
            )
        dispatcher.addReplayLocationListener(replayLocationListener)

        return dispatcher
    }

    private fun startRoute(
        point: Point,
        lastLocation: Location,
        callback: LocationEngineCallback<LocationEngineResult>
    ) {
        handler.removeCallbacks(this)
        converter.updateSpeed(speed)
        converter.updateDelay(delay)
        converter.initializeTime()
        val route = obtainRoute(point, lastLocation)
        mockedLocations =
            converter.calculateMockLocations(converter.sliceRoute(route)).toMutableList()
        dispatcher = obtainDispatcher(callback)
        dispatcher.run()
    }

    private fun obtainRoute(point: Point, lastLocation: Location): LineString {
        val pointList = ArrayList<Point>()
        pointList.add(Point.fromLngLat(lastLocation.longitude, lastLocation.latitude))
        pointList.add(point)
        return LineString.fromLngLats(pointList)
    }

    private fun scheduleNextDispatch() {
        val currentMockedPoints = mockedLocations.size
        when {
            currentMockedPoints == ZERO -> handler.postDelayed(this, DO_NOT_DELAY.toLong())
            currentMockedPoints <= MOCKED_POINTS_LEFT_THRESHOLD -> handler.postDelayed(
                this,
                ONE_SECOND_IN_MILLISECONDS.toLong()
            )
            else -> handler.postDelayed(
                this,
                ((currentMockedPoints - MOCKED_POINTS_LEFT_THRESHOLD) * ONE_SECOND_IN_MILLISECONDS).toLong()
            )
        }
    }

    private fun beginReplayWith(callback: LocationEngineCallback<LocationEngineResult>) {
        route?.let {
            start(it, callback)
        } ?: point?.let {
            startRoute(it, lastLocation, callback)
        } ?: callback.onFailure(Exception("No route found to replay."))
    }

    private fun initializeLastLocation() {
        if (!::lastLocation.isInitialized) {
            lastLocation = Location(REPLAY_ROUTE)
        }
    }
}
